ee/clickhouse/materialized_columns/columns.py

```diff
@@ -30,52 +33,116 @@
+                raise ValueError(f"got {len(columns)} columns, expected 0 or 1")
+
+
+@dataclass(frozen=True)
+class MaterializedColumnDetails:
+    table_column: TableColumn
+    property_name: PropertyName
+    is_disabled: bool
+
+    COMMENT_PREFIX = "column_materializer"
+    COMMENT_SEPARATOR = "::"
+    COMMENT_DISABLED_MARKER = "disabled"
+
+    def as_column_comment(self) -> str:
+        bits = [self.COMMENT_PREFIX, self.table_column, self.property_name]
+        if self.is_disabled:
+            bits.append(self.COMMENT_DISABLED_MARKER)
+        return self.COMMENT_SEPARATOR.join(bits)
+
+    @classmethod
+    def from_column_comment(cls, comment: str) -> MaterializedColumnDetails:
+        match comment.split(cls.COMMENT_SEPARATOR, 3):
+            # Old style comments have the format "column_materializer::property", dealing with the default table column.
+            case [cls.COMMENT_PREFIX, property_name]:
+                return MaterializedColumnDetails(DEFAULT_TABLE_COLUMN, property_name, is_disabled=False)
+            # Otherwise, it's "column_materializer::table_column::property" for columns that are active.
+            case [cls.COMMENT_PREFIX, table_column, property_name]:
+                return MaterializedColumnDetails(cast(TableColumn, table_column), property_name, is_disabled=False)
+            # Columns that are marked as disabled have an extra trailer indicating their status.
+            case [cls.COMMENT_PREFIX, table_column, property_name, cls.COMMENT_DISABLED_MARKER]:
+                return MaterializedColumnDetails(cast(TableColumn, table_column), property_name, is_disabled=True)
+            case _:
+                raise ValueError(f"unexpected comment format: {comment!r}")
+
+
 def get_materialized_columns(
     table: TablesWithMaterializedColumns,
+    exclude_disabled_columns: bool = False,
 ) -> dict[tuple[PropertyName, TableColumn], ColumnName]:
-    rows = sync_execute(
-        """
-        SELECT comment, name
-        FROM system.columns
-        WHERE database = %(database)s
-          AND table = %(table)s
-          AND comment LIKE '%%column_materializer::%%'
-          AND comment not LIKE '%%column_materializer::elements_chain::%%'
-    """,
-        {"database": CLICKHOUSE_DATABASE, "table": table},
-    )
-    if rows and get_instance_setting("MATERIALIZED_COLUMNS_ENABLED"):
-        return {_extract_property(comment): column_name for comment, column_name in rows}
-    else:
+    if not get_instance_setting("MATERIALIZED_COLUMNS_ENABLED"):
         return {}
 
+    return {
+        (column.details.property_name, column.details.table_column): column.name
+        for column in MaterializedColumn.get_all(table)
+        if not (exclude_disabled_columns and column.details.is_disabled)
+    }
+
+
+def get_on_cluster_clause_for_table(table: TableWithProperties) -> str:
+    return f"ON CLUSTER '{CLICKHOUSE_CLUSTER}'" if table == "events" else ""
```

Comment:
Why this condition? Shouldn't we always materialize on the whole cluster?

Reply:
I'm only 90% confident in my understanding of this, but as far as I understand it, it's only necessary to run the mutation once per shard. So,

1. for replicated, non-sharded tables (e.g. `persons`), this can be run on any arbitrary replica and the cluster will manage replication of any side effects, so `ON CLUSTER` isn't necessary
2. for replicated, sharded tables (e.g. `sharded_events`), this only needs to be applied to one replica in the shard -- but we don't try to be too clever with routing in this case, and just run the query `ON CLUSTER` ([note that `TableWithProperties` doesn't contain `sharded_events`](https://github.com/PostHog/posthog/blob/3183e605833243009d7a72d0f80f891d5138f60e/posthog/models/property/property.py#L47) so any queries that actually are run against `sharded_events` use the result of this function called with `events`.) I'm not sure if there are any drawbacks of doing this, @fuziontech might know
3. for distributed tables (e.g. `events`) this needs to be run on all members of the cluster, so always `ON CLUSTER`

I do think the way this was/is written is pretty confusing/not very self-explanatory  but doing a bigger refactor here for clarity's sake is something I didn't really want to tackle right now (but might look into in the future)